diff --git a/modules/imgproc/src/filter.dispatch.cpp b/modules/imgproc/src/filter.dispatch.cpp
index 3d949a27bd..0000000000 100644
--- a/modules/imgproc/src/filter.dispatch.cpp
+++ b/modules/imgproc/src/filter.dispatch.cpp
@@ -1620,12 +1620,13 @@ void filter2D(InputArray _src, OutputArray _dst, int ddepth,
     Size wsz(src.cols, src.rows);
     if( (borderType & BORDER_ISOLATED) == 0 )
         src.locateROI( wsz, ofs );
+    int borderMode = borderType & ~BORDER_ISOLATED;
 
     hal::filter2D(src.type(), dst.type(), kernel.type(),
                   src.data, src.step, dst.data, dst.step,
                   dst.cols, dst.rows, wsz.width, wsz.height, ofs.x, ofs.y,
                   kernel.data, kernel.step,  kernel.cols, kernel.rows,
                   anchor.x, anchor.y,
-                  delta, borderType, src.isSubmatrix());
+                  delta, borderMode, src.isSubmatrix());
 }
 
 void sepFilter2D(InputArray _src, OutputArray _dst, int ddepth,
